lib/autoPagination.js

```diff
@@ -0,0 +1,190 @@
+  };
+}
+
+function getDoneCallback(args) {
+  if (args.length < 2) {
+    return undefined;
+  }
+  var onDone = args[1];
+  if (typeof onDone !== 'function') {
+    throw Error('The second argument to autoPagingEach, if present, must be a callback function; receieved ' + typeof onDone);
+  }
+  return onDone;
+}
+
+function wrapAsyncIteratorWithCallback(asyncIteratorNext, onItem) {
+  return new Promise(function(resolve, reject) {
+    function handleIteration(iterResult) {
+      if (iterResult.done) {
+        resolve();
+        return;
+      }
+
+      var item = iterResult.value;
+      return new Promise(function(next) {
+        // Bit confusing, perhaps; we pass a `resolve` fn
+        // to the user, so they can decide when and if to continue.
+        onItem(item, next);
+      }).then(function(shouldContinue) {
+        if (shouldContinue === false) {
+          return handleIteration({done: true});
+        } else {
+          return asyncIteratorNext().then(handleIteration);
+        }
+      });
+    }
+
+    asyncIteratorNext().then(handleIteration).catch(reject);
+  });
+}
+
+function autoPagingEach(self, requestArgs, spec, firstPagePromise) {
+  return function autoPagingEach(/* onItem?, onDone? */) {
+    var args = [].slice.call(arguments);
+    var onItem = getItemCallback(args);
+    var onDone = getDoneCallback(args);
+    if (args.length > 2) {
+      throw Error('autoPagingEach takes up to two arguments; received:', args);
+    }
+
+    function requestNextPage(listResult) {
+      var lastIdx = listResult.data.length - 1;
+      var lastItem = listResult.data[lastIdx];
+      var lastId = lastItem && lastItem.id;
+      if (!lastId) {
+        throw Error('Unexpected: No `id` found on the last item while auto-paging a list.');
+      }
+      return makeRequest(self, requestArgs, spec, {starting_after: lastId});
+    }
+
+    // If a user calls `.next()` multiple times in parallel,
+    // return the same result until something has resolved
+    // to prevent page-turning race conditions.
+    var currentPromise;
+    function memoizedPromise(cb) {
+      if (currentPromise) {
+        return currentPromise;
+      }
+      currentPromise = new Promise(cb).then(function(ret) {
+        currentPromise = undefined;
+        return ret;
+      });
+      return currentPromise;
+    }
+
+    // Iterator state.
+    var listPromise = firstPagePromise;
+    var i = 0;
+
+    function iterate(listResult) {
+      if (!(listResult && listResult.data && typeof listResult.data.length === 'number')) {
+        throw Error('Unexpected: Stripe API response does not have a well-formed `data` array.');
+      }
+
+      if (i < listResult.data.length) {
+        var value = listResult.data[i];
+        i += 1;
+        return {value: value, done: false};
+      } else if (listResult.has_more) {
+        // Reset counter, request next page, and recurse.
+        i = 0;
+        listPromise = requestNextPage(listResult);
```

Comment:
Is there any potential harm to going to an arbitrary stack depth like this? Like if we iterated a thousand pages, do we have a thousand functions sitting on the stack and the potential to overflow? I know some languages do tail call optimization for this kind of situation, and it looks like newer versions of JS have it, but not sure if we should expect it to kick in here or not.

Reply:
Hmm, interesting. I wasn't sure about this but it looks like recursive promises are handled by the promise implementation, _not_ the call stack: 

> That recursive call is asynchronous and so it does not abuse the call stack by definition.

Which I got from [this post](https://alexn.org/blog/2017/10/11/javascript-promise-leaks-memory.html), ironically about how the promise implementation leaks memory (fortunately, at rates which don't seem problematic – I think it's not until you get well into the millions at a minimum).

So I think we're good here; thanks for checking

Reply:
K, thanks for looking into that.